File: geomesa-kafka/geomesa-kafka-datastore/src/test/scala/org/locationtech/geomesa/kafka/streams/GeoMesaStreamsBuilderTest.scala
/***********************************************************************
<<<<<<< HEAD
 * Copyright (c) 2013-2024 Commonwealth Computer Research, Inc.
=======
<<<<<<< HEAD
 * Copyright (c) 2013-2023 Commonwealth Computer Research, Inc.
=======
 * Copyright (c) 2013-2022 Commonwealth Computer Research, Inc.
>>>>>>> de758f45a6 (GEOMESA-3198 Kafka streams integration (#2854))
>>>>>>> 133afd3681 (GEOMESA-3198 Kafka streams integration (#2854))
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Apache License, Version 2.0
 * which accompanies this distribution and is available at
 * http://www.opensource.org/licenses/apache2.0.php.
 ***********************************************************************/

package org.locationtech.geomesa.kafka.streams

<<<<<<< HEAD
import org.apache.kafka.clients.consumer.{ConsumerConfig, ConsumerRecord, KafkaConsumer}
import org.apache.kafka.clients.producer.ProducerRecord
import org.apache.kafka.common.serialization.{ByteArrayDeserializer, ByteArraySerializer, LongDeserializer, StringDeserializer}
import org.apache.kafka.streams
import org.apache.kafka.streams.kstream.{Transformer, TransformerSupplier}
import org.apache.kafka.streams.processor.{ProcessorContext, WallclockTimestampExtractor}
import org.apache.kafka.streams.test.TestRecord
=======
import com.typesafe.scalalogging.StrictLogging
import org.apache.kafka.clients.consumer.{ConsumerConfig, ConsumerRecord, KafkaConsumer}
import org.apache.kafka.clients.producer.ProducerRecord
import org.apache.kafka.common.serialization.{ByteArrayDeserializer, LongDeserializer, StringDeserializer}
import org.apache.kafka.streams
import org.apache.kafka.streams.kstream.{Transformer, TransformerSupplier}
import org.apache.kafka.streams.processor.{ProcessorContext, WallclockTimestampExtractor}
>>>>>>> de758f45a6 (GEOMESA-3198 Kafka streams integration (#2854))
import org.apache.kafka.streams.{StreamsConfig, TopologyTestDriver}
import org.geotools.api.data.{DataStoreFinder, Query, Transaction}
import org.junit.runner.RunWith
import org.locationtech.geomesa.features.ScalaSimpleFeature
<<<<<<< HEAD
import org.locationtech.geomesa.kafka.KafkaContainerTest
=======
import org.locationtech.geomesa.kafka.EmbeddedKafka
>>>>>>> de758f45a6 (GEOMESA-3198 Kafka streams integration (#2854))
import org.locationtech.geomesa.kafka.data.KafkaDataStore
import org.locationtech.geomesa.utils.collection.SelfClosingIterator
import org.locationtech.geomesa.utils.geotools.converters.FastConverter
import org.locationtech.geomesa.utils.geotools.{FeatureUtils, SimpleFeatureTypes}
import org.locationtech.geomesa.utils.io.WithClose
<<<<<<< HEAD
=======
import org.specs2.mutable.Specification
>>>>>>> de758f45a6 (GEOMESA-3198 Kafka streams integration (#2854))
import org.specs2.runner.JUnitRunner

import java.nio.charset.StandardCharsets
import java.time.Duration
import java.util.concurrent.ConcurrentHashMap
import java.util.{Collections, Date, Properties}
import scala.collection.mutable.ArrayBuffer
import scala.concurrent.duration.DurationInt

@RunWith(classOf[JUnitRunner])
<<<<<<< HEAD
class GeoMesaStreamsBuilderTest extends KafkaContainerTest {

  import org.apache.kafka.streams.scala.ImplicitConversions._
  import org.apache.kafka.streams.scala.kstream._
  import org.apache.kafka.streams.scala.serialization.Serdes._
=======
class GeoMesaStreamsBuilderTest extends Specification with StrictLogging {

  import org.apache.kafka.streams.scala.ImplicitConversions._
  import org.apache.kafka.streams.scala.Serdes._
  import org.apache.kafka.streams.scala.kstream._
>>>>>>> de758f45a6 (GEOMESA-3198 Kafka streams integration (#2854))

  import scala.collection.JavaConverters._

  lazy val sft =
    SimpleFeatureTypes.createImmutableType("streams", "name:String,age:Int,dtg:Date,*geom:Point:srid=4326")

  lazy val features = Seq.tabulate(10) { i =>
    ScalaSimpleFeature.create(sft, s"id$i", s"name$i", i % 2, s"2022-04-27T00:00:0$i.00Z", s"POINT(1 $i)")
  }

<<<<<<< HEAD
=======
  var kafka: EmbeddedKafka = _

  step {
    logger.info("Starting embedded kafka/zk")
    kafka = new EmbeddedKafka()
    logger.info("Started embedded kafka/zk")
  }

>>>>>>> de758f45a6 (GEOMESA-3198 Kafka streams integration (#2854))
  private val zkPaths = Collections.newSetFromMap(new ConcurrentHashMap[String, java.lang.Boolean]())

  def getParams(zkPath: String): Map[String, String] = {
    require(zkPaths.add(zkPath), s"zk path '$zkPath' is reused between tests, may cause conflicts")
    Map(
<<<<<<< HEAD
      "kafka.brokers"            -> brokers,
      "kafka.zookeepers"         -> zookeepers,
=======
      "kafka.brokers"            -> kafka.brokers,
      "kafka.zookeepers"         -> kafka.zookeepers,
>>>>>>> de758f45a6 (GEOMESA-3198 Kafka streams integration (#2854))
      "kafka.topic.partitions"   -> "1",
      "kafka.topic.replication"  -> "1",
      "kafka.consumer.read-back" -> "Inf",
      "kafka.zk.path"            -> zkPath
    )
  }

  "GeoMesaStreamsBuilder" should {
    "read from geomesa topics" in {
      val params = getParams("word/count")
      // write features to the embedded kafka
      val kryoTopic = WithClose(DataStoreFinder.getDataStore(params.asJava)) { ds =>
        ds.createSchema(sft)
        WithClose(ds.getFeatureWriterAppend(sft.getTypeName, Transaction.AUTO_COMMIT)) { writer =>
          features.foreach(FeatureUtils.write(writer, _, useProvidedFid = true))
        }
        KafkaDataStore.topic(ds.getSchema(sft.getTypeName))
      }

      // read off the features from kafka
      val messages: Seq[ConsumerRecord[Array[Byte], Array[Byte]]] = {
        val props = new Properties()
<<<<<<< HEAD
        props.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, brokers)
=======
        props.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, kafka.brokers)
>>>>>>> de758f45a6 (GEOMESA-3198 Kafka streams integration (#2854))
        props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest")
        props.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, classOf[ByteArrayDeserializer].getName)
        props.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, classOf[ByteArrayDeserializer].getName)
        props.put(ConsumerConfig.GROUP_ID_CONFIG, "consume-kryo-topic")
        val buf = ArrayBuffer.empty[ConsumerRecord[Array[Byte], Array[Byte]]]
        WithClose(new KafkaConsumer[Array[Byte], Array[Byte]](props)) { consumer =>
          consumer.subscribe(Collections.singleton(kryoTopic))
          val start = System.currentTimeMillis()
          while (buf.lengthCompare(10) < 0 && System.currentTimeMillis() - start < 10000) {
            consumer.poll(Duration.ofMillis(100)).asScala.foreach { record =>
              buf += record
            }
          }
        }
        buf.toSeq
<<<<<<< HEAD
=======
        buf
<<<<<<< HEAD
>>>>>>> de758f45a6 (GEOMESA-3198 Kafka streams integration (#2854))
=======
>>>>>>> de758f45a (GEOMESA-3198 Kafka streams integration (#2854))
>>>>>>> 1b8cbf843d (GEOMESA-3198 Kafka streams integration (#2854))
=======
>>>>>>> d845d7c1bd (GEOMESA-3254 Add Bloop build support)
      }

      val timestampExtractor = new TimestampExtractingTransformer()

      val builder = GeoMesaStreamsBuilder(params)

      // pull out the timestamps with an identity transform for later comparison
      val streamFeatures = builder.stream(sft.getTypeName).transform[String, GeoMesaMessage](
        new TransformerSupplier[String, GeoMesaMessage, streams.KeyValue[String, GeoMesaMessage]]() {
          override def get(): Transformer[String, GeoMesaMessage, streams.KeyValue[String, GeoMesaMessage]] =
            timestampExtractor
        })

      // word count example copied from https://kafka.apache.org/31/documentation/streams/developer-guide/dsl-api.html#scala-dsl
      val textLines: KStream[String, String] =
        streamFeatures.map((k, v) => (k, v.attributes.map(_.toString.replaceAll(" ", "_")).mkString(" ")))
      val wordCounts: KTable[String, Long] =
        textLines
            .flatMapValues(textLine => textLine.split(" +"))
            .groupBy((_, word) => word)
            .count()(Materialized.as("counts-store"))

      wordCounts.toStream.to("word-count")

      val props = new Properties()
      props.put(StreamsConfig.APPLICATION_ID_CONFIG, "word-count-test")
<<<<<<< HEAD
=======
      props.put(StreamsConfig.APPLICATION_ID_CONFIG, "geomesa-test-app")
<<<<<<< HEAD
>>>>>>> de758f45a6 (GEOMESA-3198 Kafka streams integration (#2854))
=======
>>>>>>> de758f45a (GEOMESA-3198 Kafka streams integration (#2854))
>>>>>>> 1b8cbf843d (GEOMESA-3198 Kafka streams integration (#2854))
=======
>>>>>>> d845d7c1bd (GEOMESA-3254 Add Bloop build support)
      props.put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, "dummy:1234")

      val output = scala.collection.mutable.Map.empty[String, java.lang.Long]
      WithClose(new TopologyTestDriver(builder.build(), props)) { testDriver =>
<<<<<<< HEAD
        val inputTopic = testDriver.createInputTopic(kryoTopic, new ByteArraySerializer(), new ByteArraySerializer())
        messages.foreach(m => inputTopic.pipeInput(new TestRecord[Array[Byte],Array[Byte]](m)))
        val outputTopic = testDriver.createOutputTopic("word-count", new StringDeserializer(), new LongDeserializer())
        while (!outputTopic.isEmpty) {
          val rec = outputTopic.readRecord
          output += rec.key() -> rec.value()
=======
        messages.foreach(testDriver.pipeInput)
        var out = testDriver.readOutput("word-count", new StringDeserializer(), new LongDeserializer())
        while (out != null) {
          output += out.key() -> out.value()
          out = testDriver.readOutput("word-count", new StringDeserializer(), new LongDeserializer())
>>>>>>> de758f45a6 (GEOMESA-3198 Kafka streams integration (#2854))
        }
      }

      val expected =
        features.flatMap(_.getAttributes.asScala.map(_.toString.replaceAll(" ", "_")))
            .groupBy(s => s)
            .map { case (k, v) => k -> v.length.toLong }

      output mustEqual expected

      val expectedTimestamps = features.map(_.getAttribute("dtg").asInstanceOf[Date].getTime)
      foreach(timestampExtractor.timestamps)(_._2 must haveLength(1))
      timestampExtractor.timestamps.map(_._2.head) must containTheSameElementsAs(expectedTimestamps)
    }

    "write to geomesa topics" in {
      val params = getParams("write/test")
      // create the output feature type and topic
      val kryoTopic = WithClose(DataStoreFinder.getDataStore(params.asJava)) { ds =>
        ds.createSchema(sft)
        KafkaDataStore.topic(ds.getSchema(sft.getTypeName))
      }

      val testInput = features.map { sf =>
        val offset = sf.getID.replace("id", "").toLong
        val key = sf.getID.getBytes(StandardCharsets.UTF_8)
        val value = sf.getAttributes.asScala.map(FastConverter.convert(_, classOf[String])).mkString(",")
        new ConsumerRecord("input-topic", 0, offset, key, value.getBytes(StandardCharsets.UTF_8))
      }

      val builder = GeoMesaStreamsBuilder(params)

      val input: KStream[String, String] =
        builder.wrapped.stream[String, String]("input-topic")(implicitly[Consumed[String, String]].withTimestampExtractor(new WallclockTimestampExtractor()))
      val output: KStream[String, GeoMesaMessage] =
        input.mapValues { lines =>
          GeoMesaMessage.upsert(lines.split(","))
        }

      builder.to(sft.getTypeName, output)

      val props = new Properties()
      props.put(StreamsConfig.APPLICATION_ID_CONFIG, "write-test")
<<<<<<< HEAD
=======
      props.put(StreamsConfig.APPLICATION_ID_CONFIG, "geomesa-test-app")
<<<<<<< HEAD
>>>>>>> de758f45a6 (GEOMESA-3198 Kafka streams integration (#2854))
=======
>>>>>>> de758f45a (GEOMESA-3198 Kafka streams integration (#2854))
>>>>>>> 1b8cbf843d (GEOMESA-3198 Kafka streams integration (#2854))
=======
>>>>>>> d845d7c1bd (GEOMESA-3254 Add Bloop build support)
      props.put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, "dummy:1234")

      val kryoMessages = ArrayBuffer.empty[ProducerRecord[Array[Byte], Array[Byte]]]
      WithClose(new TopologyTestDriver(builder.build(), props)) { testDriver =>
<<<<<<< HEAD
        val inputTopic = testDriver.createInputTopic("input-topic", new ByteArraySerializer(), new ByteArraySerializer())
        testInput.foreach(m => inputTopic.pipeInput(new TestRecord[Array[Byte],Array[Byte]](m)))
        val outputTopic = testDriver.createOutputTopic(kryoTopic, new ByteArrayDeserializer(), new ByteArrayDeserializer())
        while (!outputTopic.isEmpty) {
          val rec = outputTopic.readRecord
          kryoMessages += new ProducerRecord(kryoTopic, 0, rec.timestamp, rec.getKey, rec.getValue, rec.getHeaders)
=======
        testInput.foreach(testDriver.pipeInput)
        var out = testDriver.readOutput(kryoTopic, new ByteArrayDeserializer(), new ByteArrayDeserializer())
        while (out != null) {
          kryoMessages += out
          out = testDriver.readOutput(kryoTopic, new ByteArrayDeserializer(), new ByteArrayDeserializer())
>>>>>>> de758f45a6 (GEOMESA-3198 Kafka streams integration (#2854))
        }
      }

      WithClose(DataStoreFinder.getDataStore(params.asJava)) { case ds: KafkaDataStore =>
        // initialize kafka consumers for the store
        ds.getFeatureReader(new Query(sft.getTypeName), Transaction.AUTO_COMMIT).close()
        // send the mocked messages to the actual embedded kafka topic
        WithClose(KafkaDataStore.producer(ds.config.brokers, Map.empty)) { producer =>
          kryoMessages.foreach(producer.send)
        }
        eventually(40, 100.millis) {
          val result = SelfClosingIterator(ds.getFeatureReader(new Query(sft.getTypeName), Transaction.AUTO_COMMIT)).toList
          result must containTheSameElementsAs(features)
        }
      }
    }
  }

<<<<<<< HEAD
=======
  step {
    logger.info("Stopping embedded kafka/zk")
    kafka.close()
    logger.info("Stopped embedded kafka/zk")
  }

>>>>>>> de758f45a6 (GEOMESA-3198 Kafka streams integration (#2854))
  class TimestampExtractingTransformer
      extends Transformer[String, GeoMesaMessage, org.apache.kafka.streams.KeyValue[String, GeoMesaMessage]]() {

    private var context: ProcessorContext = _

    val timestamps = scala.collection.mutable.Map.empty[String, ArrayBuffer[Long]]

    override def init(context: ProcessorContext): Unit = this.context = context

    override def transform(key: String, value: GeoMesaMessage): streams.KeyValue[String, GeoMesaMessage] = {
      timestamps.getOrElseUpdate(key, ArrayBuffer.empty) += context.timestamp()
      new streams.KeyValue(key, value)
    }

    override def close(): Unit = {}
  }

}

File: geomesa-kafka/geomesa-kafka-datastore/src/test/java/org/locationtech/geomesa/kafka/jstreams/GeoMesaStreamsBuilderTest.java
/***********************************************************************
<<<<<<< HEAD
 * Copyright (c) 2013-2024 Commonwealth Computer Research, Inc.
=======
<<<<<<< HEAD
 * Copyright (c) 2013-2023 Commonwealth Computer Research, Inc.
=======
 * Copyright (c) 2013-2022 Commonwealth Computer Research, Inc.
>>>>>>> 58d14a257e (GEOMESA-3254 Add Bloop build support)
>>>>>>> 1463162d60 (GEOMESA-3254 Add Bloop build support)
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Apache License, Version 2.0
 * which accompanies this distribution and is available at
 * http://www.opensource.org/licenses/apache2.0.php.
 ***********************************************************************/

package org.locationtech.geomesa.kafka.jstreams;

import org.apache.kafka.clients.consumer.ConsumerConfig;
import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.apache.kafka.clients.consumer.KafkaConsumer;
import org.apache.kafka.clients.producer.Producer;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.apache.kafka.common.serialization.ByteArrayDeserializer;
<<<<<<< HEAD
import org.apache.kafka.common.serialization.ByteArraySerializer;
import org.apache.kafka.common.serialization.LongDeserializer;
import org.apache.kafka.common.serialization.Serdes;
import org.apache.kafka.common.serialization.Serdes.StringSerde;
import org.apache.kafka.common.serialization.StringDeserializer;
import org.apache.kafka.streams.KeyValue;
import org.apache.kafka.streams.StreamsConfig;
import org.apache.kafka.streams.TestInputTopic;
import org.apache.kafka.streams.TestOutputTopic;
=======
import org.apache.kafka.common.serialization.LongDeserializer;
import org.apache.kafka.common.serialization.Serdes;
import org.apache.kafka.common.serialization.StringDeserializer;
import org.apache.kafka.streams.KeyValue;
import org.apache.kafka.streams.StreamsConfig;
>>>>>>> 58d14a257e (GEOMESA-3254 Add Bloop build support)
import org.apache.kafka.streams.TopologyTestDriver;
import org.apache.kafka.streams.kstream.Consumed;
import org.apache.kafka.streams.kstream.KStream;
import org.apache.kafka.streams.kstream.KTable;
import org.apache.kafka.streams.kstream.Materialized;
import org.apache.kafka.streams.kstream.Produced;
import org.apache.kafka.streams.kstream.Transformer;
import org.apache.kafka.streams.processor.ProcessorContext;
import org.apache.kafka.streams.processor.WallclockTimestampExtractor;
<<<<<<< HEAD
import org.apache.kafka.streams.test.TestRecord;
<<<<<<< HEAD
import org.geotools.api.data.DataStoreFinder;
import org.geotools.api.data.Query;
import org.geotools.api.data.SimpleFeatureReader;
import org.geotools.api.data.SimpleFeatureWriter;
import org.geotools.api.data.Transaction;
import org.geotools.api.feature.simple.SimpleFeature;
import org.geotools.api.feature.simple.SimpleFeatureType;
=======
=======
>>>>>>> 58d14a257e (GEOMESA-3254 Add Bloop build support)
import org.geotools.data.DataStoreFinder;
import org.geotools.data.Query;
import org.geotools.data.Transaction;
import org.geotools.data.simple.SimpleFeatureReader;
import org.geotools.data.simple.SimpleFeatureWriter;
>>>>>>> 4a4bbd8ec03 (GEOMESA-3254 Add Bloop build support)
import org.junit.AfterClass;
import org.junit.Assert;
import org.junit.BeforeClass;
import org.junit.Test;
import org.locationtech.geomesa.features.ScalaSimpleFeature;
<<<<<<< HEAD
=======
import org.locationtech.geomesa.kafka.EmbeddedKafka;
>>>>>>> 58d14a257e (GEOMESA-3254 Add Bloop build support)
import org.locationtech.geomesa.kafka.data.KafkaDataStore;
import org.locationtech.geomesa.kafka.streams.GeoMesaMessage;
import org.locationtech.geomesa.utils.geotools.FeatureUtils;
import org.locationtech.geomesa.utils.geotools.SimpleFeatureTypes;
import org.locationtech.geomesa.utils.geotools.converters.FastConverter;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
<<<<<<< HEAD
import org.testcontainers.containers.KafkaContainer;
import org.testcontainers.containers.output.Slf4jLogConsumer;
import org.testcontainers.utility.DockerImageName;
=======
>>>>>>> 58d14a257e (GEOMESA-3254 Add Bloop build support)

import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Properties;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;

public class GeoMesaStreamsBuilderTest {

    private static final Logger logger = LoggerFactory.getLogger(GeoMesaStreamsBuilderTest.class);

<<<<<<< HEAD
    static KafkaContainer container = null;
=======
    static EmbeddedKafka kafka = null;
>>>>>>> 58d14a257e (GEOMESA-3254 Add Bloop build support)

    static final SimpleFeatureType sft =
          SimpleFeatureTypes.createImmutableType("streams", "name:String,age:Int,dtg:Date,*geom:Point:srid=4326");

    static final List<SimpleFeature> features = new ArrayList<>();

    static final Set<String> zkPaths = Collections.newSetFromMap(new ConcurrentHashMap<>());

<<<<<<< HEAD
    static String zookeepers() {
        return String.format("%s:%s", container.getHost(), container.getMappedPort(KafkaContainer.ZOOKEEPER_PORT));
    }
    static String brokers() {
        return container.getBootstrapServers();
    }

=======
>>>>>>> 58d14a257e (GEOMESA-3254 Add Bloop build support)
    public Map<String, String> getParams(String zkPath) {
        if (!zkPaths.add(zkPath)) {
            throw new IllegalArgumentException("zk path '" + zkPath + "' is reused between tests, may cause conflicts");
        }
        Map<String, String> params = new HashMap<>();
<<<<<<< HEAD
        params.put("kafka.brokers", brokers());
        params.put("kafka.zookeepers", zookeepers());
=======
        params.put("kafka.brokers", kafka.brokers());
        params.put("kafka.zookeepers", kafka.zookeepers());
>>>>>>> 58d14a257e (GEOMESA-3254 Add Bloop build support)
        params.put("kafka.topic.partitions", "1");
        params.put("kafka.topic.replication", "1");
        params.put("kafka.consumer.read-back", "Inf");
        params.put("kafka.zk.path", zkPath);
        return params;
    }

    @BeforeClass
    public static void init() {
<<<<<<< HEAD
        DockerImageName image =
              DockerImageName.parse("confluentinc/cp-kafka")
                             .withTag(System.getProperty("confluent.docker.tag", "7.3.1"));
        container = new KafkaContainer(image);
        container.start();
        container.followOutput(new Slf4jLogConsumer(logger));
=======
        logger.info("Starting embedded kafka/zk");
        kafka = new EmbeddedKafka();
        logger.info("Started embedded kafka/zk");
>>>>>>> 58d14a257e (GEOMESA-3254 Add Bloop build support)

        for (int i = 0; i < 10; i ++) {
            ScalaSimpleFeature sf = new ScalaSimpleFeature(sft, "id" + i, null, null);
            sf.setAttribute(0, "name" + i);
            sf.setAttribute(1, i % 2);
            sf.setAttribute(2, "2022-04-27T00:00:0" + i + ".00Z");
            sf.setAttribute(3, "POINT(1 " + i + ")");
            features.add(sf);
        }
    }

    @AfterClass
    public static void destroy() {
<<<<<<< HEAD
        if (container != null) {
            container.stop();
        }
=======
        logger.info("Stopping embedded kafka/zk");
        if (kafka != null) {
            kafka.close();
        }
        logger.info("Stopped embedded kafka/zk");
>>>>>>> 58d14a257e (GEOMESA-3254 Add Bloop build support)
    }

    @Test
    public void testRead() throws Exception {
        Map<String, String> params = getParams("word/count");
        String kryoTopic;

        // write features to the embedded kafka
        KafkaDataStore ds = (KafkaDataStore) DataStoreFinder.getDataStore(params);
        try {
            ds.createSchema(sft);
            try (SimpleFeatureWriter writer = ds.getFeatureWriterAppend(sft.getTypeName(), Transaction.AUTO_COMMIT)) {
                features.forEach(f -> FeatureUtils.write(writer, f, true));
            }
            kryoTopic = KafkaDataStore.topic(ds.getSchema(sft.getTypeName()));
        } finally {
            ds.dispose();
        }

        Properties consumerProps = new Properties();
<<<<<<< HEAD
        consumerProps.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, brokers());
=======
        consumerProps.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, kafka.brokers());
>>>>>>> 58d14a257e (GEOMESA-3254 Add Bloop build support)
        consumerProps.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
        consumerProps.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, ByteArrayDeserializer.class.getName());
        consumerProps.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, ByteArrayDeserializer.class.getName());
        consumerProps.put(ConsumerConfig.GROUP_ID_CONFIG, "consume-kryo-topic");

        // read off the features from kafka
        List<ConsumerRecord<byte[], byte[]>> messages = new ArrayList<>();
        try (KafkaConsumer<byte[], byte[]> consumer = new KafkaConsumer<>(consumerProps)) {
            consumer.subscribe(Collections.singleton(kryoTopic));
            long start = System.currentTimeMillis();
            while (messages.size() < 10 && System.currentTimeMillis() - start < 10000) {
                consumer.poll(Duration.ofMillis(100)).forEach(messages::add);
            }
        }

        TimestampExtractingTransformer timestampExtractor = new TimestampExtractingTransformer();

        GeoMesaStreamsBuilder builder = GeoMesaStreamsBuilder.create(params);

        // pull out the timestamps with an identity transform for later comparison
        KStream<String, GeoMesaMessage> streamFeatures =
              builder.stream(sft.getTypeName()).transform(() -> timestampExtractor);

        // word count example copied from https://kafka.apache.org/31/documentation/streams/developer-guide/dsl-api.html#scala-dsl
        KStream<String, String> textLines = streamFeatures.mapValues(v ->
            v.asJava().stream().map(a -> a.toString().replaceAll(" ", "_")).collect(Collectors.joining(" ")));
        KTable<String, Long> wordCounts = textLines
              .flatMapValues(textLine -> Arrays.asList(textLine.split(" +")))
              .groupBy((k, word) -> word)
              .count(Materialized.as("counts-store"));
        wordCounts.toStream().to("word-count", Produced.with(Serdes.String(), Serdes.Long()));

        Properties streamsProps = new Properties();
        streamsProps.put(StreamsConfig.APPLICATION_ID_CONFIG, "java-word-count-test");
        streamsProps.put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, "dummy:1234");
<<<<<<< HEAD
        streamsProps.put(StreamsConfig.DEFAULT_KEY_SERDE_CLASS_CONFIG, StringSerde.class);
        streamsProps.put(StreamsConfig.DEFAULT_VALUE_SERDE_CLASS_CONFIG, StringSerde.class);

        Map<String, Long> output;
        try (TopologyTestDriver testDriver = new TopologyTestDriver(builder.build(), streamsProps)) {
            TestInputTopic<byte[], byte[]> inputTopic = testDriver.createInputTopic(kryoTopic,
                                                                                    new ByteArraySerializer(),
                                                                                    new ByteArraySerializer());
            messages.forEach(m -> inputTopic.pipeInput(new TestRecord<>(m)));
            TestOutputTopic<String, Long> outputTopic = testDriver.createOutputTopic("word-count", new StringDeserializer(), new LongDeserializer());
            output = new HashMap<>(outputTopic.readKeyValuesToMap());
=======
        streamsProps.put(StreamsConfig.DEFAULT_KEY_SERDE_CLASS_CONFIG, Serdes.String().getClass());
        streamsProps.put(StreamsConfig.DEFAULT_VALUE_SERDE_CLASS_CONFIG, Serdes.String().getClass());

        Map<String, Long> output = new HashMap<>();
        try (TopologyTestDriver testDriver = new TopologyTestDriver(builder.build(), streamsProps)) {
            messages.forEach(testDriver::pipeInput);
            ProducerRecord<String, Long> out =
                  testDriver.readOutput("word-count", new StringDeserializer(), new LongDeserializer());
            while (out != null) {
                output.put(out.key(), out.value());
                out = testDriver.readOutput("word-count", new StringDeserializer(), new LongDeserializer());
            }
>>>>>>> 58d14a257e (GEOMESA-3254 Add Bloop build support)
        }

        Map<String, Long> expected = new HashMap<>();
        features.stream().flatMap(f -> f.getAttributes().stream().map(a -> a.toString().replaceAll(" ", "_")))
                .forEach(w -> expected.compute(w, (word, count) -> count == null ? 1L : count + 1));

        Assert.assertEquals(expected, output);
        List<Long> expectedTimestamps =
              features.stream().map(f -> ((Date) f.getAttribute("dtg")).getTime()).collect(Collectors.toList());
        timestampExtractor.timestamps.values().forEach(c -> Assert.assertEquals(1, c.size()));
        List<Long> timestamps =
              timestampExtractor.timestamps.values()
                                           .stream()
                                           .map(c -> c.get(0))
                                           .sorted()
                                           .collect(Collectors.toList());
        Assert.assertEquals(expectedTimestamps, timestamps);
    }

    @Test
    public void testWrite() throws Exception {
        Map<String, String> params = getParams("write/test");
        String kryoTopic;

        // create the output feature type and topic
        KafkaDataStore ds = (KafkaDataStore) DataStoreFinder.getDataStore(params);
        try {
            ds.createSchema(sft);
            kryoTopic = KafkaDataStore.topic(ds.getSchema(sft.getTypeName()));
        } finally {
            ds.dispose();
        }

        List<ConsumerRecord<byte[], byte[]>> testInput = features.stream().map(sf -> {
            long offset = Long.parseLong(sf.getID().replace("id", ""));
            byte[] key = sf.getID().getBytes(StandardCharsets.UTF_8);
            String value =
                  sf.getAttributes()
                    .stream()
                    .map(a -> FastConverter.convert(a, String.class))
                    .collect(Collectors.joining(","));
            return new ConsumerRecord<>("input-topic", 0, offset, key, value.getBytes(StandardCharsets.UTF_8));
        }).collect(Collectors.toList());

        GeoMesaStreamsBuilder builder = GeoMesaStreamsBuilder.create(params);

        KStream<String, String> input =
            builder.wrapped().stream("input-topic",
                 Consumed.with(Serdes.String(), Serdes.String()).withTimestampExtractor(new WallclockTimestampExtractor()));
        KStream<String, GeoMesaMessage> output =
<<<<<<< HEAD
            input.mapValues(lines -> GeoMesaMessage.upsert(Arrays.asList((Object[])lines.split(","))));
=======
            input.mapValues(lines -> GeoMesaMessage.upsert(Arrays.asList(lines.split(","))));
>>>>>>> 58d14a257e (GEOMESA-3254 Add Bloop build support)

        builder.to(sft.getTypeName(), output);

        Properties streamsProps = new Properties();
        streamsProps.put(StreamsConfig.APPLICATION_ID_CONFIG, "java-write-test");
        streamsProps.put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, "dummy:1234");
<<<<<<< HEAD
        streamsProps.put(StreamsConfig.DEFAULT_KEY_SERDE_CLASS_CONFIG, StringSerde.class);
        streamsProps.put(StreamsConfig.DEFAULT_VALUE_SERDE_CLASS_CONFIG, StringSerde.class);

        List<ProducerRecord<byte[], byte[]>> kryoMessages = new ArrayList<>();
        try (TopologyTestDriver testDriver = new TopologyTestDriver(builder.build(), streamsProps)) {
            TestInputTopic<byte[], byte[]> inputTopic = testDriver.createInputTopic("input-topic",
                                                                                    new ByteArraySerializer(),
                                                                                    new ByteArraySerializer());
            testInput.forEach(m -> inputTopic.pipeInput(new TestRecord<>(m)));
            TestOutputTopic<byte[], byte[]> outputTopic = testDriver.createOutputTopic(kryoTopic, new ByteArrayDeserializer(), new ByteArrayDeserializer());
            while (!outputTopic.isEmpty()) {
                TestRecord<byte[], byte[]> rec = outputTopic.readRecord();
                kryoMessages.add(new ProducerRecord<>(kryoTopic, 0, rec.timestamp(), rec.getKey(), rec.getValue(), rec.getHeaders()));
=======
        streamsProps.put(StreamsConfig.DEFAULT_KEY_SERDE_CLASS_CONFIG, Serdes.String().getClass());
        streamsProps.put(StreamsConfig.DEFAULT_VALUE_SERDE_CLASS_CONFIG, Serdes.String().getClass());

        List<ProducerRecord<byte[], byte[]>> kryoMessages = new ArrayList<>();
        try (TopologyTestDriver testDriver = new TopologyTestDriver(builder.build(), streamsProps)) {
            testInput.forEach(testDriver::pipeInput);
            ProducerRecord<byte[], byte[]> out =
                  testDriver.readOutput(kryoTopic, new ByteArrayDeserializer(), new ByteArrayDeserializer());
            while (out != null) {
                kryoMessages.add(out);
                out = testDriver.readOutput(kryoTopic, new ByteArrayDeserializer(), new ByteArrayDeserializer());
>>>>>>> 58d14a257e (GEOMESA-3254 Add Bloop build support)
            }
        }

        List<SimpleFeature> result = new ArrayList<>();
        ds = (KafkaDataStore) DataStoreFinder.getDataStore(params);
        try {
            // initialize kafka consumers for the store
            ds.getFeatureReader(new Query(sft.getTypeName()), Transaction.AUTO_COMMIT).close();
            // send the mocked messages to the actual embedded kafka topic
            try(Producer<byte[], byte[]> producer = KafkaDataStore.producer(ds.config())) {
                kryoMessages.forEach(producer::send);
            }

            long end = System.currentTimeMillis() + 4000;
            while (System.currentTimeMillis() < end) {
                try (SimpleFeatureReader reader = ds.getFeatureReader(new Query(sft.getTypeName()), Transaction.AUTO_COMMIT)) {
                    while (reader.hasNext()) {
                        result.add(reader.next());
                    }
                }
                result.sort(Comparator.comparing(SimpleFeature::getID));
                if (result.equals(features)) {
                    break;
                } else {
                    result.clear();
                    Thread.sleep(100);
                }
            }
        } finally {
            ds.dispose();
        }
        Assert.assertEquals(features, result);
    }

    static class TimestampExtractingTransformer
          implements Transformer<String, GeoMesaMessage, KeyValue<String, GeoMesaMessage>> {

        private ProcessorContext context = null;

        Map<String, List<Long>> timestamps = new HashMap<>();

        @Override
        public void init(ProcessorContext context) {
            this.context = context;
        }

        @Override
        public KeyValue<String, GeoMesaMessage> transform(String key, GeoMesaMessage value) {
            timestamps.computeIfAbsent(key, k -> new ArrayList<>()).add(context.timestamp());
            return new KeyValue<>(key, value);
        }

        @Override
        public void close() {

        }
    }
}
